sources/api/netdog/src/networkd_status.rs

```diff
@@ -0,0 +1,309 @@
+//! The networkd_status module contains definitions and functions for tracking network status from networkd
+//!
+//! The primary purpose of this module is to provide structures to read in a `networkctl status` command output
+//! and make specific fields available from this output. These structs can then be used to read DNS, IP Addressing
+//! and any other networking status data for use in configuration files needed for other networking tools.
+use crate::interface_id::InterfaceName;
+use crate::NETWORKCTL;
+use serde::de::Error;
+use serde::{Deserialize, Deserializer};
+use snafu::{ensure, ResultExt};
+use std::cmp::Ordering;
+use std::convert::TryInto;
+use std::net::IpAddr;
+use std::process::Command;
+
+#[derive(Clone, Debug, Deserialize)]
+#[serde(rename_all = "PascalCase")]
+pub(crate) struct NetworkdInterfaceStatus {
+    pub(crate) name: InterfaceName,
+    #[serde(rename = "DNS")]
+    pub(crate) dns: Option<Vec<NetworkdDnsConfig>>,
+    pub(crate) search_domains: Option<Vec<SearchDomain>>,
+    #[serde(rename = "Addresses", deserialize_with = "from_networkctl_addresses")]
+    pub(crate) addresses: Vec<IpAddr>,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub(crate) struct NetworkdDnsConfig {
+    #[serde(rename = "Address", deserialize_with = "ipaddr_from_vec_de")]
+    pub(crate) address: IpAddr,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+#[serde(deny_unknown_fields)]
+#[allow(dead_code)]
+pub(crate) struct SearchDomain {
+    #[serde(rename = "Domain")]
+    pub(crate) domain: String,
+    #[serde(rename = "ConfigSource")]
+    config_source: String,
+    #[serde(rename = "ConfigProvider", deserialize_with = "ipaddr_from_vec_de")]
+    config_provider: IpAddr,
+}
+
+// get an IpAddr from a Vec<u8> (could be 4 or 16 length)
+fn ipaddr_from_vec(address_vec: Vec<u8>) -> Result<IpAddr> {
+    match address_vec.len() {
+        4 => {
+            // We checked that its exactly 4 u8 long, this should not be possible to fail, yet the compiler isn't certain
+            let octets: [u8; 4] = address_vec
+                .try_into()
+                .expect("unable to parse IP Address from Vec");
```

Comment:
There has been a lot of various back and forth on this issue, I was able to avoid `panic` explicitly with another match being explicit on the `std::result::Result` coming back from this `into` logic. As an aside, we also toyed with code similar to:
```rust
let octets: [u8; 4] = [
    *address_vec.get(0).context(error::BadIpAddressVecSnafu)?,
    *address_vec.get(1).context(error::BadIpAddressVecSnafu)?,
    *address_vec.get(2).context(error::BadIpAddressVecSnafu)?,
    *address_vec.get(3).context(error::BadIpAddressVecSnafu)?, 
];
 Ok(IpAddr::from(octets))
```
This is very specific and protected by the match but looks much more verbose for IPv6 which needs 16 of these. Ultimately we need to pick what way we like best and just go with it unless someone finds a better way to get from the `Vec<u8>` to the `[u8; 4]` needed by `IpAddr::from()`'s interface.